front: add support for string unions for keys in i18n-checker

emersion opened the pull request:
See https://github.com/emersion/i18next-typescript-parser/pull/2

emersion committed:
front: invert conditional for step.stopType

TypeScript isn't smart enough to figure out that the stopType
cannot be undefined in the else branch.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: add support for string unions for keys in i18n-checker

See https://github.com/emersion/i18next-typescript-parser/pull/2

Signed-off-by: Simon Ser <contact@emersion.fr>

Synar commented on front/src/applications/stdcm/components/StdcmResults/SimulationReportSheet.tsx:
```diff
@@ -446,9 +446,9 @@ const SimulationReportSheet = ({
-                            : capitalizeFirstLetter(t(`trainPath.stopType.${step.stopType}`))}
+                          {step.stopType
+                            ? capitalizeFirstLetter(t(`trainPath.stopType.${step.stopType}`))
+                            : t('reportSheet.serviceStop')}
```
Older "issue" than this pr, but this can be simplified, as first and last step now have step.stopType = serviceStop. So we can skip the ternary.

emersion replied:
Hm I'd prefer to keep this for a separate PR, this doesn't seem immediately obvious to me.

Synar reviewed (approved):
I've added a comment for a small simplification we can do but other than that LGTM, thanks!

I'm a bit mystified though by how well this works for template strings according to your tests in (https://github.com/emersion/i18next-typescript-parser/pull/2). If I try 
```typescript
  const getChoice = () => (Math.random() > 0.5 ? 'h' : 'i');
  const a = `${getChoice()}`;
  const b: 'h' | 'i' = a;
  console.log(b);
  ```
  I will have typescript scream at me that I can't convert from string to 'h' | 'i'. How come in the AST you seem to get a union type from the template and not just a string? Is it first parsed as an union and devolved to a string at a later point? Or am I missing something?

emersion commented:
> I will have typescript scream at me that I can't convert from string to 'h' | 'i'. How come in the AST you seem to get a union type from the template and not just a string? Is it first parsed as an union and devolved to a string at a later point? Or am I missing something?

By default, TypeScript will use the more general type available, but it's also possible to narrow the type as much as possible with `as const`:

```ts
  const getChoice = () => (Math.random() > 0.5 ? 'h' : 'i');
  const a = `${getChoice()}` as const;
  const b: 'h' | 'i' = a;
  console.log(b);
```

See also: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#const-type-parameters

Synar commented:
Thanks!

But since you don't use as const in your tests, does this mean that checker.getTypeAtLocation retrieve the narrowest type by default, without the need to use const type or as const?

Synar commented on front/scripts/i18n-checker.ts:
```diff
@@ -222,10 +221,23 @@ function visitCallExpression(
   const optionsNode = node.arguments.length > 1 ? node.arguments.at(-1) : undefined;
 
   // Extract the key from the first function argument
```
You should probably update this comment ^^

emersion replied:
Done!

emersion commented:
`TFunction` is defined with a `const Key` generic type parameter, so will use the narrowest type:

https://github.com/i18next/i18next/blob/defa00f1a3bc79a32645d95f737b5ce4dcace14b/typescript/t.d.ts#L307

Synar commented:
> `TFunction` is defined with a `const Key` generic type parameter, so will use the narrowest type:
> 
> https://github.com/i18next/i18next/blob/defa00f1a3bc79a32645d95f737b5ce4dcace14b/typescript/t.d.ts#L307

Oh nice, that does make our life easier! Thanks for the explanation!

emersion committed:
front: handle keyPrefix TFunction option in i18n-checker

See:

- https://github.com/i18next/i18next-gitbook/pull/234
- https://github.com/emersion/i18next-typescript-parser/commit/b9d4484d22025efa315f5889cc8582aa61ecfd25

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: remove unnecessary cast in splitRollingStockProperties()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: narrow down labels in RollingStockEditorParameterFormColumn

SchemaProperty contains two kinds of properties: parameters and
metadata. This component only deals with parameters.

Makes the i18n-checker script happier.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: fix missing i18n key nesting in RollingStockEditorCategoryForm

Due to ignoreJSONStructure, i18next will try to look deeply into
the translation file (in any child object) if a translation is
missing. As a result the translation shows up even if the key is
incomplete.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: drop getObjTypeAction() and getObjTypeEdition()

TypeScript isn't smart enough to figure out that objectTypeEdition
and objectTypeAction are tied together here. This upsets
i18n-checker because translation keys are constructed from these
variables.

Drop the trivial function and inline the logic with a single
conditional.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: print source file location for missing i18n keys

Makes it easier to figure out where a key is used, especially for
dynamic keys.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: remove unused type in DateBox props

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: fix again null category not handled in the occurrence indicator

97f3ec3e576d ("front: fix null category not handled in the occurrence
indicator") fixed this already, but this got lost in
aa5ce9590092 ("front: add exception").

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: add missing translation for main.timetable.invalid.zero_length_path

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: fix typo in manageTrainSchedule.pathfindingErrors.invalid_path_items

"items" is pluralized in the error code.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: add missing manageTrainSchedule.pathfindingErrors.zero_length_path

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: cut down IGNORE_UNUSED list in i18n-checker

Now that our script supports TypeScript unions, we can remove a
whole bunch of these.

More work is needed to drop the rest.

Signed-off-by: Simon Ser <contact@emersion.fr>

SharglutDev reviewed (approved):
Great job, lgtm (trusting you @Synar for the tech part :sweat_smile: )
